sycl/test/basic_tests/half_type.cpp

```diff
@@ -103,17 +103,95 @@ void verify_vec(queue &q) {
   assert(e.get_access<access::mode::read>()[0] == 0);
 }
 
+void verify_numeric_limits(queue &q) {
+  // Verify on host side
+  // Static member variables
+  std::numeric_limits<cl::sycl::half>::is_specialized;
+  std::numeric_limits<cl::sycl::half>::is_signed;
+  std::numeric_limits<cl::sycl::half>::is_integer;
+  std::numeric_limits<cl::sycl::half>::is_exact;
+  std::numeric_limits<cl::sycl::half>::has_infinity;
+  std::numeric_limits<cl::sycl::half>::has_quiet_NaN;
+  std::numeric_limits<cl::sycl::half>::has_signaling_NaN;
+  std::numeric_limits<cl::sycl::half>::has_denorm;
+  std::numeric_limits<cl::sycl::half>::has_denorm_loss;
+  std::numeric_limits<cl::sycl::half>::tinyness_before;
+  std::numeric_limits<cl::sycl::half>::traps;
+  std::numeric_limits<cl::sycl::half>::max_exponent10;
+  std::numeric_limits<cl::sycl::half>::max_exponent;
+  std::numeric_limits<cl::sycl::half>::min_exponent10;
+  std::numeric_limits<cl::sycl::half>::min_exponent;
+  std::numeric_limits<cl::sycl::half>::radix;
+  std::numeric_limits<cl::sycl::half>::max_digits10;
+  std::numeric_limits<cl::sycl::half>::digits;
+  std::numeric_limits<cl::sycl::half>::is_bounded;
+  std::numeric_limits<cl::sycl::half>::digits10;
+  std::numeric_limits<cl::sycl::half>::is_modulo;
+  std::numeric_limits<cl::sycl::half>::is_iec559;
+  std::numeric_limits<cl::sycl::half>::round_style;
+
+  // Static member functions
+  std::numeric_limits<cl::sycl::half>::min();
+  std::numeric_limits<cl::sycl::half>::max();
+  std::numeric_limits<cl::sycl::half>::lowest();
+  std::numeric_limits<cl::sycl::half>::epsilon();
+  std::numeric_limits<cl::sycl::half>::round_error();
+  std::numeric_limits<cl::sycl::half>::infinity();
+  std::numeric_limits<cl::sycl::half>::quiet_NaN();
+  std::numeric_limits<cl::sycl::half>::signaling_NaN();
+  std::numeric_limits<cl::sycl::half>::denorm_min();
+
+  // Verify in kernel function for device side check
+  q.submit([&](cl::sycl::handler &cgh) {
+    // Static member variables
+    std::numeric_limits<cl::sycl::half>::is_specialized;
```

Comment:
Shame on me…Why could the above part check on device side as well?

Reply:
Both host and device compilers parses the whole source file so missing class member would be catch during that process. Device code is generated only for the kernel function and its dependencies, so as long as don't need to execute the code to perform the checks, keeping only that part of the code is enough.
But I've just realized that it would better to keep instead the part inside the kernel function: it's compiled by both host and device compilers and also executed on OpenCL device and host device as well. In such way we would also verify that there are no issues during some additional device specific processing with numeric_limits class.

Reply:
Yep, this is my initial expectation. :)

Reply:
This is a great benefit of the the review process! :-)

Reply:
Yeah, it's true. Thank you guys very much! :)